### ISSUE TITLE
Regression issue for use id_a:empty string and id_a_size:0 to sm2 sign in OpenSSL 3.0.2

### ISSUE THREAD

**Issue by Wenxing-hou** [openssl/openssl#18022]
When we use **id_a: empty sting and id_a_size:0** to sm2 sign from OpenSSL 3.0.2, it failed. And it is passed in OpenSSL 1.1.1 before.
```
bool libspdm_sm2_dsa_sign(const void *sm2_context, uintn hash_nid,
                          const uint8_t *id_a, uintn id_a_size,
                          const uint8_t *message, uintn size,
                          uint8_t *signature, uintn *sig_size)
```
We refer to SPDM spec version [1.2.0](https://www.dmtf.org/sites/default/files/standards/documents/DSP0274_1.2.0.pdf). For the line 835:  _**Lastly, SM2 expects a distinguishing identifier, which identifies the signer, and is indicated by the variable IDA. If this algorithm is selected, the ID shall be an empty string of size 0.**_

I  find the patch **_params: avoid a core dump with a null pointer and a get string call #11474_** changed it.  

***
**The overall code flow is:**

```
libspdm_sm2_dsa_sign
|
V
pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);       //init pkey_ctx null
|
V
int EVP_PKEY_CTX_set1_id(EVP_PKEY_CTX *ctx, const void *id, int len) 
        if (data_len > 0) {
            ctx->cached_parameters.dist_id = OPENSSL_memdup(data, data_len);  //when data_len is 0, dist_id is still null
            if (ctx->cached_parameters.dist_id == NULL) {
                ERR_raise(ERR_LIB_EVP, ERR_R_MALLOC_FAILURE);
                return 0;
            }
        }
        ctx->cached_parameters.dist_id_set = 1;
        ctx->cached_parameters.dist_id_len = data_len;
|
V
result = EVP_DigestSignInit(ctx, NULL, EVP_sm3(), NULL, pkey);   // p->data = ctx->cached_parameters.dist_id
|
V
    if (p->data == NULL)
        return 0;                         //The final error place:   p->data == NULL
```  
***
The detail code flow is:
```
bool libspdm_sm2_dsa_sign(const void *sm2_context, uintn hash_nid,
                          const uint8_t *id_a, uintn id_a_size,
                          const uint8_t *message, uintn size,
                          uint8_t *signature, uintn *sig_size)
|
V
pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);         //init pkey_ctx 
|
V
result = EVP_PKEY_CTX_set1_id(pkey_ctx, id_a,   //id_a is empty string; id_a_size is 0
                                  id_a_size);
|
V
int EVP_PKEY_CTX_set1_id(EVP_PKEY_CTX *ctx, const void *id, int len)
{
    return EVP_PKEY_CTX_ctrl(ctx, -1, -1,
                             EVP_PKEY_CTRL_SET1_ID, (int)len, (void*)(id));
}
|
V
    ret = evp_pkey_ctx_store_cached_data(ctx, keytype, optype,
                                         cmd, NULL, p2, p1);
|
V
        if (data_len > 0) {
            ctx->cached_parameters.dist_id = OPENSSL_memdup(data, data_len);
            if (ctx->cached_parameters.dist_id == NULL) {
                ERR_raise(ERR_LIB_EVP, ERR_R_MALLOC_FAILURE);
                return 0;
            }
        }
        ctx->cached_parameters.dist_id_set = 1;
        ctx->cached_parameters.dist_id_len = data_len;
|
V
result = EVP_DigestSignInit(ctx, NULL, EVP_sm3(), NULL, pkey);
|
V
return do_sigver_init(ctx, pctx, type, NULL, NULL, NULL, e, pkey, 0,NULL)  
|
V
ret = signature->digest_sign_init(locpctx->op.sig.algctx,
                                          mdname, provkey, params);
|
V
static int sm2sig_digest_signverify_init(void *vpsm2ctx, const char *mdname,
                                         void *ec, const OSSL_PARAM params[])

            *ctx->params =
                OSSL_PARAM_construct_octet_string(translation->param_key,
                                                  ctx->p2, (size_t)ctx->p1); //init params
|
V
    if (ret > 0)
        ret = evp_pkey_ctx_use_cached_data(locpctx);
|
v
    if (ret && ctx->cached_parameters.dist_id_set) {
        const char *name = ctx->cached_parameters.dist_id_name;
        const void *val = ctx->cached_parameters.dist_id;
        size_t len = ctx->cached_parameters.dist_id_len;

            ret = evp_pkey_ctx_ctrl_int(ctx, -1, ctx->operation,
                                        EVP_PKEY_CTRL_SET1_ID,
                                        (int)len, (void *)val);   //len is 0
    }
|
V
    switch (evp_pkey_ctx_state(ctx)) {
    case EVP_PKEY_STATE_PROVIDER:
        return evp_pkey_ctx_ctrl_to_param(ctx, keytype, optype, cmd, p1, p2);
|
V
static int default_fixup_args(enum state state,
                              const struct translation_st *translation,
                              struct translation_ctx_st *ctx)
|
V
static int default_check(enum state state,
                         const struct translation_st *translation,
                         const struct translation_ctx_st *ctx)    
|
V
        case SET:
            ret = evp_pkey_ctx_set_params_strict(pctx, ctx.params);
            break;
|
V
    return EVP_PKEY_CTX_set_params(ctx, params);
|
V
            return
                ctx->op.sig.signature->set_ctx_params(ctx->op.sig.algctx,
                                                      params);
|
V
static int sm2sig_set_ctx_params(void *vpsm2ctx, const OSSL_PARAM params[])
|
V
        p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_DIST_ID);    // gen p
|
V
OSSL_PARAM *OSSL_PARAM_locate(OSSL_PARAM *p, const char *key) //p = params,and p->data = null
{
    if (p != NULL && key != NULL)
        for (; p->key != NULL; p++)
            if (strcmp(key, p->key) == 0)
                return p;                                                
    return NULL;
}
|
V
        if (!OSSL_PARAM_get_octet_string(p, &tmp_id, 0, &tmp_idlen))
            return 0;
|
V
static int get_string_internal(const OSSL_PARAM *p, void **val,
                               size_t *max_len, size_t *used_len,
                               unsigned int type)

    if (p->data == NULL)
        return 0;                         //The final error place:   p->data == NULL
```

**Comment by mattcaswell** [comment]
What happens if you just don't call `EVP_PKEY_CTX_set1_id`? It seems it doesn't support setting a zero length id, but the default is zero length anyway - so it should just work without this.

**Comment by Wenxing-hou** [comment]
Thanks. I have fixed the issue with the  [commit ID](https://github.com/DMTF/libspdm/commit/1329601211299d05ad6c082403a1b44104ec6de7).
It is really help me. I think we can close this issue.

**Comment by t8m** [comment]
IMO it is fine that there exist a workaround but it is a regression. The provider should handle the case with 0 length id data correctly.

**Comment by t8m** [comment]
Fix in #18052
